orderer/consensus/etcdraft/util.go

```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
+			if err != nil {
+				return nil, errors.Wrap(err, "parsing tls root certs")
+			}
+			intermediateCerts, err := parseCertificateListFromBytes(org.MSP().GetTLSIntermediateCerts())
+			if err != nil {
+				return nil, errors.Wrap(err, "parsing tls intermediate certs")
+			}
+
+			for _, cert := range rootCerts {
+				fingerprint, err := getCertFingerprint(cert)
+				if err != nil {
+					return nil, errors.WithStack(err)
+				}
+
+				if _, exists := tlsRootsMap[fingerprint]; !exists {
+					tlsRoots.AddCert(cert)
+					tlsRootsMap[fingerprint] = cert
+				}
+			}
+
+			for _, cert := range intermediateCerts {
+				fingerprint, err := getCertFingerprint(cert)
+				if err != nil {
+					return nil, errors.WithStack(err)
+				}
+
+				if _, exists := tlsIntermediatesMap[fingerprint]; !exists {
+					tlsIntermediates.AddCert(cert)
+					tlsIntermediatesMap[fingerprint] = cert
+				}
+			}
+		}
 	}
 
-	if _, err := x509.ParseCertificate(bl.Bytes); err != nil {
-		return errors.Errorf("%s TLS certificate has invalid ASN1 structure, %v: %s", certRole, err, string(pemData))
+	if newOrdererConfig != nil {
+		for _, org := range newOrdererConfig.Organizations() {
+			rootCerts, err := parseCertificateListFromBytes(org.MSP().GetTLSRootCerts())
+			if err != nil {
+				return nil, errors.Wrap(err, "parsing tls root certs")
+			}
+			intermediateCerts, err := parseCertificateListFromBytes(org.MSP().GetTLSIntermediateCerts())
+			if err != nil {
+				return nil, errors.Wrap(err, "parsing tls intermediate certs")
+			}
+
+			for _, cert := range rootCerts {
+				fingerprint, err := getCertFingerprint(cert)
+				if err != nil {
+					return nil, errors.WithStack(err)
+				}
+
+				if _, exists := tlsRootsMap[fingerprint]; !exists {
+					tlsRoots.AddCert(cert)
+					tlsRootsMap[fingerprint] = cert
+				}
+			}
+
+			for _, cert := range intermediateCerts {
+				fingerprint, err := getCertFingerprint(cert)
+				if err != nil {
+					return nil, errors.WithStack(err)
+				}
+
+				if _, exists := tlsIntermediatesMap[fingerprint]; !exists {
+					tlsIntermediates.AddCert(cert)
+					tlsIntermediatesMap[fingerprint] = cert
+				}
+			}
+		}
 	}
+
+	return &x509.VerifyOptions{
+		Roots:         tlsRoots,
+		Intermediates: tlsIntermediates,
+		KeyUsages: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageClientAuth,
+			x509.ExtKeyUsageServerAuth,
+		},
+	}, nil
+}
+
+//validateConsenterTLSCerts decodes PEM cert, parses and validates it. If opts is nil, then only sanity check is done.
+func validateConsenterTLSCerts(c *etcdraft.Consenter, opts *x509.VerifyOptions, ignoreExpiration bool) error {
+	clientCert, err := parseCertificateFromBytes(c.ClientTlsCert)
+	if err != nil {
+		return errors.Wrapf(err, "parsing tls client cert of %s:%d", c.Host, c.Port)
+	}
+
+	serverCert, err := parseCertificateFromBytes(c.ServerTlsCert)
+	if err != nil {
+		return errors.Wrapf(err, "parsing tls server cert of %s:%d", c.Host, c.Port)
+	}
+
+	if opts != nil {
```

Comment:
If we have nil options, then aren't we badly screwed?

Reply:
`validateConsenterTLSCerts` could is used only for sanity check(parse x509) if `opts` is nil.

Reply:
fixed.